game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/EvaderRetreat.java

```diff
@@ -0,0 +1,143 @@
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+import lombok.Builder;
+import lombok.experimental.UtilityClass;
+import org.triplea.sound.SoundUtils;
+
+@UtilityClass
+public class EvaderRetreat {
+
+  @Builder(toBuilder = true)
+  static class Parameters {
+    BattleState battleState;
+    BattleActions battleActions;
+    BattleState.Side side;
+    IDelegateBridge bridge;
+    Collection<Territory> possibleRetreatSites;
+    Collection<Unit> units;
+    String step;
+  }
+
+  public static void retreatUnits(final Parameters parameters) {
+    final BattleState battleState = parameters.battleState;
+    final BattleActions battleActions = parameters.battleActions;
+    final BattleState.Side side = parameters.side;
+    final IDelegateBridge bridge = parameters.bridge;
+    final Collection<Territory> possibleRetreatSites = parameters.possibleRetreatSites;
+    final Collection<Unit> units = parameters.units;
+    final String step = parameters.step;
+    final GamePlayer retreatingPlayer =
+        side == BattleState.Side.DEFENSE ? battleState.getDefender() : battleState.getAttacker();
+    final String text = retreatingPlayer.getName() + " retreat subs?";
+
+    bridge.getDisplayChannelBroadcaster().gotoBattleStep(battleState.getBattleId(), step);
+    final boolean isAttemptingSubmerge =
+        possibleRetreatSites.size() == 1
+            && possibleRetreatSites.contains(battleState.getBattleSite());
+    final Territory retreatTo =
+        isAttemptingSubmerge
+            ? battleActions.querySubmergeTerritory(
+                battleState, bridge, retreatingPlayer, possibleRetreatSites, text)
+            : battleActions.queryRetreatTerritory(
+                battleState, bridge, retreatingPlayer, possibleRetreatSites, text);
+    if (retreatTo == null) {
+      return;
+    }
+    playSound(battleState, bridge, retreatingPlayer, units);
+    if (battleState.getBattleSite().equals(retreatTo)) {
+      submergeEvaders(battleState, battleActions, units, side, bridge);
+      broadcastRetreat(bridge, retreatingPlayer, step, " submerges subs");
+    } else {
+
+      final CompositeChange change = new CompositeChange();
+      change.add(ChangeFactory.moveUnits(battleState.getBattleSite(), retreatTo, units));
+      bridge.addChange(change);
+      battleState.retreatUnits(side, units);
+
+      addHistoryRetreat(bridge, units, " retreated to " + retreatTo.getName());
+      notifyRetreat(battleState, battleActions, units, side, bridge);
+      broadcastRetreat(
+          bridge, retreatingPlayer, step, " retreats", " retreats subs to " + retreatTo.getName());
+    }
+  }
+
+  private static void playSound(
```

Comment:
A method for a single line method call seems a overkill. Is the 'isHeadless' check truly needed? Even if it is, what do you think about inlining this?